### PR TITLE
A PR to address the bug in numba binning

### PR DESCRIPTION
**Pull request by rettigl**
See #66 for details.
This initial commit contains additional test data to reproduce and test for the bug.

### CODE REVIEW

**Comment by rettigl** [comment]
This PR addresses two bugs in numba binning from bins ranges:
- samples on the last bin edge are discarded, while they are included in the last bin in np.histogram and in numba from bins. Specific tests for this have been implemented, and these cases have been included.
- under rare circumstances, bin edges are slightly off due to numerical rounding errors. A test for such a case has been included, and a bufgix has been implemented by rounding bin numbers to 11 digits of precision. While this fixes the bug I have identified, and the other random tests do pass, I cannot guarantee there are no further cases where this might still fail. However, I find theres edge cases also not so relevant, as no events are getting lost, but just shifted between bins, which is anyways irrelevant when jittering is employed.

Performace is lightly affected, but only by a few percent, see benchmarks in #66 and the corresponding benchmark notebook, which is also added here.

File: sed/binning/numba_bin.py [hunk 1]
```diff
@@ -55,7 +55,11 @@ def _hist_from_bin_range(
         is_inside = True
         flatidx = 0
         for i in range(ndims):
-            j = (sample[t, i] - ranges[i, 0]) * delta[i]
+            # strip off numerical rounding errors
+            j = round((sample[t, i] - ranges[i, 0]) * delta[i], 11)
+            # add counts on last edge
+            if j == bins[i]:
```
Comment (daviddoji) [inline sed/binning/numba_bin.py:61]: instead of comparing, what about something like this:
```
j = j - 0.5 * (j == bins[i])
```
the operation is then vectorized and we don't have to compare in every step

File: tests/test_binning.py [hunk 2]
```diff
@@ -17,17 +17,35 @@
...
+
+sample_int = np.random.randint(
+    low=60000,
+    high=70001,
+    size=(int(1e5), 3),
+    dtype=int,
+)
+bins_int = tuple(np.random.randint(10, 300, size=3, dtype=int))
+ranges_int = [
+    tuple(np.sort(np.random.randint(60000, 70000, size=2, dtype=int))),
```
Comment (daviddoji) [inline tests/test_binning.py:38]: why repeating the same tuple 3 times?

File: sed/binning/numba_bin.py [hunk 3]
```diff
@@ -55,7 +55,11 @@ def _hist_from_bin_range(
         is_inside = True
         flatidx = 0
         for i in range(ndims):
-            j = (sample[t, i] - ranges[i, 0]) * delta[i]
+            # strip off numerical rounding errors
+            j = round((sample[t, i] - ranges[i, 0]) * delta[i], 11)
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/binning/numba_bin.py:59]: I'm not sure I understand what you mean. We do create the array in advance (line 45). The round here happens to strip off numerical round errors that happen for certain combinations of edge values. 
The underlying reason is that here we do the forward transformation (position -> bin index), while in np.histogramdd linspace is used to create the bin edges, i.e. (bin index -> position). These two directions can produce slightly different results due to numerical float rounding errors, which I strip off at the 11th digit with this round command, which emperically does the trick, at least for the test case I identified.

File: sed/binning/numba_bin.py (hunk 1 above)
Comment (rettigl) [reply sed/binning/numba_bin.py:61]: We loop through each event, and each dimension individually (for t, for dim), so j is a single float number. I don't think that this will make a difference.

File: sed/binning/numba_bin.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (daviddoji) [reply sed/binning/numba_bin.py:59]: > I'm not sure I understand what you mean. We do create the array in advance (line 45). The round here happens to strip off numerical round errors that happen for certain combinations of edge values. The underlying reason is that here we do the forward transformation (position -> bin index), while in np.histogramdd linspace is used to create the bin edges, i.e. (bin index -> position). These two directions can produce slightly different results due to numerical float rounding errors, which I strip off at the 11th digit with this round command, which emperically does the trick, at least for the test case I identified.

Sorry, my bad. I read that `j` was an array instead of a single float. As I said, as `round` is now supported in numba, the change makes sense.

File: sed/binning/numba_bin.py (hunk 1 above)
Comment (daviddoji) [reply sed/binning/numba_bin.py:61]: Same as above.

File: tests/test_binning.py (hunk 2 above)
Comment (daviddoji) [reply tests/test_binning.py:38]: I understand the code. My point was about repeating three times the same line. Why not something like:
```
ranges_int = [tuple(np.sort(np.random.randint(60000, 70000, size=2, dtype=int))) for _ in range(3)]
```

Comment (rettigl) [reply tests/test_binning.py:38]: I see, this is probably the pythonic way of writing this. I personally find it a bit more difficult to understand, but I can change it.

**Comment by daviddoji** [comment]
LGTM!

**Comment by rettigl** [comment]
> LGTM!

Can you confirm the PR? Then I don't have to overwrite branch protection.
